[Console] Argument consumption

Guikingone opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #23343
| License       | MIT
| Doc PR        | 

Hi everyone, 

This PR is linked to #23343, as explained by @nicolas-grekas, a new argument has been added in order to remove the parameter if needed. 

Tests should be completed (in progress), same things for the code (not every case are covered).

fabpot committed:
updated CHANGELOG for 3.3.16

weaverryan committed:
Adding session strategy to ALL listeners to avoid *any* possible
fixation

xabbuh committed:
clear CSRF tokens when the user is logged out

Guikingone committed:
fix(component): Console -> argument consumption

A new argument has been added in order to consume or not the argument.

Guikingone committed:
tests(component): Console -> tests started
The tests have been started

Small fix on interface implementation.

Guikingone committed:
tests(component): Console -> progress on parameter consumption

Guikingone committed:
style(component): Console -> fix on test style for dataprovider

Guikingone committed:
fix(component): Console -> fix on post-rebase

The tests have been fixed.

Same things for the ArgvInput.

Guikingone committed:
style(component): Console -> style fixed

Guikingone committed:
tests(component): Console -> improvement on foreach usage

The key is now used.

nicolas-grekas commented on src/Symfony/Component/Console/Input/ArgvInput.php:
```diff
@@ -294,7 +294,7 @@ public function hasParameterOption($values, $onlyParams = false)
      * {@inheritdoc}
      */
-    public function getParameterOption($values, $default = false, $onlyParams = false)
+    public function getParameterOption($values, $default = false, $onlyParams = false, $consume = false)
```
Only this class should be changed: all the other changes should be reverted as they are unneeded BC breaks.
On this class, the argument should be commented, mentionned in the docblock and accessed using func_get_arg().
See eg https://github.com/symfony/symfony/pull/30212/files#diff-94f4ef2d14dbabcc4f9c798fea31defbR350

Guikingone committed:
fix(component): Console -> fix on consumed argument

The consumed argument is now optional (BC promise).

Guikingone committed:
style(component): Console -> fix on count ternary

Guikingone committed:
style(component): Console -> fix on count usage

Guikingone committed:
style(component): Console -> fix on func_get_args usage

Guikingone committed:
fix(component): Console -> fix on code style & tests

The code style has been fixed, consumption to define.

Small fix on the dataprovider during tests.

Guikingone committed:
tests(component): Console -> progress on dataprovider

The dataprovider has been updated.

The code should be finished.

Guikingone committed:
fix(component): Console -> fix on parameter consumption

nicolas-grekas commented on src/Symfony/Component/Console/Tests/Input/ArgvInputTest.php:
```diff
@@ -455,4 +455,33 @@ public function testParseOptionWithValueOptionalGivenEmptyAndOptionalArgument()
+            [['app/console', 'foo:bar', '-e', 'dev'], '-e', false, false, 'dev'],
+            [['app/console', 'foo:bar', '-e', 'dev'], '-e', false, true, ''],
+            [['app/console', 'foo:bar', '-e', 'dev', '-f', 'bar'], ['-e', '-f'], false, false, 'dev'],
+            [['app/console', 'foo:bar', '-e', 'dev', '-f', 'bar'], ['-e', '-f'], false, true, ''],
```
when consume is true, the first call should return true, and the next ones should return false, because the first call "consumed" the info

nicolas-grekas commented on src/Symfony/Component/Console/Tests/Input/ArgvInputTest.php:
```diff
@@ -455,4 +455,33 @@ public function testParseOptionWithValueOptionalGivenEmptyAndOptionalArgument()
+            [['app/console', 'foo:bar', '-e', 'dev', '--foo', 'bar'], '-f', false, false, ''],
+            [['app/console', 'foo:bar', '-e', 'dev', '--foo', 'bar'], '-f', false, true, ''],
+            [['app/console', 'foo:bar', '-e', 'dev', '--foo', 'bar'], '--foo', false, false, 'bar'],
+            [['app/console', 'foo:bar', '-e', 'dev', '--foo', 'bar'], '-f', false, true, ''],
```
there are move ways to set the dev env: `--env=dev` and `--env dev` - we should test them all, not only `-e dev`

nicolas-grekas reviewed (commented):
thanks, some comments to help move forward.

nicolas-grekas commented:
please rebase also, there are unrelated commit in the current history.

nicolas-grekas commented:
Closing as the current approach is not the correct one and this is stalling, I'll take over, thanks for trying!
